src/Query/QueryBuilder.php

```diff
@@ -302,14 +309,51 @@ public function fetchFirstColumn(): array
      */
     public function executeQuery(): Result
     {
+        [$params, $types] = $this->boundParameters();
+
         return $this->connection->executeQuery(
             $this->getSQL(),
-            $this->params,
-            $this->types,
+            $params,
+            $types,
             $this->resultCacheProfile,
         );
     }
 
+    /**
+     * Retrieve parameters and types bound to all queries (optional CTEs and main query).
+     *
+     * @return array{
+     *     list<mixed>|array<string, mixed>,
+     *     WrapperParameterTypeArray,
+     * } The parameters and types bound to the CTE queries merged with those bound to the main query.
+     */
+    private function boundParameters(): array
+    {
+        if (count($this->withParts) === 0) {
+            return [$this->params, $this->types];
+        }
+
+        $cteParams = $cteParamsTypes = [];
+
+        foreach ($this->withParts as $withPart) {
+            if (! $withPart->query instanceof self) {
+                continue;
+            }
+
+            $cteParams      = array_merge($cteParams, $withPart->query->params);
```

Comment:
Why don't we stick to a consistent terminology in the query builder? Right now, "CTE" and "with" are used interchangeably. I propose that we use CTE everywhere. WITH is just a SQL keywords used to declare a block of CTEs.

The `WithSQLBuilder` name is fine since it's exactly about the syntax.

Reply:
Ok, I just wan't to be sure to understand before start changes.
Using CTE everywhere means replacing `with` by `CTE` everywhere.

What do you think about the name of the method in the QueryBuilder ?
```php

/** @var CommonTableExpression[] */
private array $commonTableExpressions = []; // or $CTEs, $cteParts?

public function with(string $name, string|QueryBuilder $query, ?array $columns = null): self
```

Should I rename also the method ?

```php
public function commonTableExpression(string $name, string|QueryBuilder $query, ?array $columns = null): self
```

Or ?

```php
public function cte(string $name, string|QueryBuilder $query, ?array $columns = null): self
```

- `commonTableExpression` is long but we understand easily what it is for
- `cte` seems odd

Reply:
```php
/** @var CommonTableExpression[] */
private array $commonTableExpressions = []; // or $CTEs, $cteParts?

public function with(string $name, string|QueryBuilder $expression, ?array $columns = null): self
```

I think this would make sense. The query operates common table expressions. The method is named `with()` to mimic the SQL syntax.